python/ray/data/tests/test_iceberg.py

```diff
@@ -293,6 +294,508 @@ def test_write_concurrency():
     assert df["col_a"].tolist() == [1, 2, 3, 4]
 
 
+@pytest.mark.skipif(
+    get_pyarrow_version() < parse_version("14.0.0"),
+    reason="PyIceberg 0.7.0 fails on pyarrow <= 14.0.0",
+)
+def test_predicate_pushdown():
+    """Test that predicate pushdown works correctly with Iceberg datasource."""
+    import pandas as pd
+
+    from ray.data._internal.logical.optimizers import LogicalOptimizer
+    from ray.data.expressions import col
+
+    # Read the table and apply filters using Ray Data expressions
+    ds = ray.data.read_iceberg(
+        table_identifier=f"{_DB_NAME}.{_TABLE_NAME}",
+        catalog_kwargs=_CATALOG_KWARGS.copy(),
+    )
+
+    # Apply filter using Ray Data expression syntax
+    filtered_ds = ds.filter(expr=col("col_c") >= 5)
+
+    # Verify the filter is pushed down to the read operation
+    # by checking the optimized logical plan
+    logical_plan = filtered_ds._plan._logical_plan
+    optimized_plan = LogicalOptimizer().optimize(logical_plan)
+    actual_plan = optimized_plan.dag.dag_str
+
+    # The plan should only contain the Read operator, with no Filter operator
+    # This indicates the filter was pushed down to the datasource
+    assert (
+        "Filter[Filter" not in actual_plan
+    ), f"Filter should be pushed down to read, got plan: {actual_plan}"
+    assert (
+        "Read[ReadIceberg]" in actual_plan
+    ), f"Expected ReadIceberg in plan, got: {actual_plan}"
+
+    # Verify the results are correct
+    result = filtered_ds.to_pandas()
+    assert all(result["col_c"] >= 5), "All rows should have col_c >= 5"
+    assert len(result) > 0, "Should have some results"
+
+    # Verify against direct PyIceberg read with the same filter
+    sql_catalog = pyi_catalog.load_catalog(**_CATALOG_KWARGS)
+    expected_table = (
+        sql_catalog.load_table(f"{_DB_NAME}.{_TABLE_NAME}")
+        .scan(row_filter=pyi_expr.GreaterThanOrEqual("col_c", 5))
+        .to_pandas()
+        .sort_values(["col_a", "col_b", "col_c"])
+        .reset_index(drop=True)
+    )
+
+    result_sorted = result.sort_values(["col_a", "col_b", "col_c"]).reset_index(
+        drop=True
+    )
+    pd.testing.assert_frame_equal(result_sorted, expected_table)
```

Comment:
(Here and elsewhere) Optionally, you can use this utility function: https://github.com/goutamvenkat-anyscale/ray/blob/24d72104ff474044d71c7be7ca06a250af968620/python/ray/data/_internal/util.py#L1670-L1689